File: 383.赎金信.py
#
# @lc app=leetcode.cn id=383 lang=python
#
# [383] 赎金信
#

# @lc code=start
class Solution(object):
    def canConstruct(self, ransomNote, magazine):
        """
        :type ransomNote: str
        :type magazine: str
        :rtype: bool
        """
        if not ransomNote:
            return True
        cnt = {}
        for c in ransomNote:
            cnt[c] = cnt.get(c, 0) + 1
        num = len(ransomNote)
        for c in magazine:
            if c in cnt and cnt[c] >= 1:
                cnt[c] -= 1
                num -= 1
            if num == 0:
                return True
        return False
# @lc code=end

File: 383.赎金信.cpp
/*
 * @lc app=leetcode.cn id=383 lang=cpp
 *
 * [383] 赎金信
 */

// @lc code=start
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<char, int> dic;
        for(auto& ch : magazine)
        {
            dic[ch] += 1;
        }
        for(auto&ch : ransomNote)
        {
            dic[ch] -= 1;
            if(dic[ch] < 0)
                return false;
        }
        return true;

    }
};
// @lc code=end

